### PR TITLE
Refactor to ignore a failure for the grace period

### PR DESCRIPTION
**Pull request by gregcorbett**
Supersedes #18, ~~depends on #23 and #22~~. Resolves GT-235.

- currently, an error in the pipeline immediately causes the next icinga check to raise an error.
- this led to flapping hour to hour, generating emails that were ignored.
- this should mean we only get alerts if the failover pipeline has not succeeded within the grace period.
- adding an output on a successful restore in 1_runDbUpdate.sh simplifies the logic needed by this check. We don't have to worry about "if/when the pipeline failed?", we only care about "when did it last succeed?".

This PR will also add:
- unit tests for the check_db_dump_recent.py script
- a GitHub workflow to run those tests at PR time
- a dependabot config to keep workflows/actions up to date

### CODE REVIEW

**Comment by gregcorbett** [comment]
The new unit tests do pass: https://github.com/gregcorbett/gocdb-failover-scripts/actions/runs/6036355188

File: .github/workflows/unit-test.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+jobs:
+  unit-test:
+    name: Python ${{ matrix.python-version }} ${{ matrix.os }} test
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        # 3.6 is the version readily available on EL7. Requries ubuntu-20.04.
```
Comment (rowan04) [inline .github/workflows/unit-test.yml:13]:
```suggestion
        # 3.6 is the version readily available on EL7. Requires ubuntu-20.04.
```

File: check/check_db_dump_recent.py [hunk 2]
```diff
@@ -23,6 +23,9 @@ def __init__(self, grace_period):
         # successful run was this long ago.
         self.GRACE_PERIOD = grace_period
 
+        # The string the database update script outputs on a sucessful restore.
```
Comment (rowan04) [inline check/check_db_dump_recent.py:26]:
```suggestion
        # The string the database update script outputs on a successful restore.
```

File: check/check_db_dump_recent.py [hunk 3]
```diff
@@ -43,17 +45,30 @@ def run(self, log_file_path):
...
 
-            # Convert the last line of the log to a timestamp.
-            # Use a inner try...except block to handle the somewhat more
-            # expected error of "the failover process ran, but something
-            # went wrong" differently.
-            try:
-                last_success = datetime.strptime(
-                    last_line,
-                    "%Y-%m-%dT%H:%M:%S%z",
+            # Assume the failover processs has never run, then attempt to
```
Comment (rowan04) [inline check/check_db_dump_recent.py:48]:
```suggestion
            # Assume the failover process has never run, then attempt to
```

**Review by rowan04 — COMMENTED** [review]
Some grammatical changes. Rest of the code lgtm

**Comment by tofu-rocketry** [comment]
Could you add some notes to the PR about the GitHub workflow changes? (Assuming they are intentionally part of this PR.)

**Comment by gregcorbett** [comment]
Done

File: importDBdmpFile/1_runDbUpdate.sh [hunk 4]
```diff
@@ -173,4 +173,5 @@ cd ..
 # Do not remove following line - another process relies on this exact string
 # being the last line in the log file
 logger "completed ok"
+echo "$(date --iso-8601='seconds') INFO: completed ok"
```
Comment (tofu-rocketry) [inline importDBdmpFile/1_runDbUpdate.sh:176]: It looks a bit weird having this `echo` in the middle of this section bookended by comments about `logger`. Could it go before it?

File: check/check_db_dump_recent.py [hunk 5]
```diff
@@ -31,9 +34,8 @@ def run(self, log_file_path):
             # Use a inner try...except block to handle the more expected error
             # of "Couldn't open/read the provided file" differently.
             try:
-                with open(log_file_path, 'r') as log_file_path:
-                    lines = log_file_path.read().splitlines()
-                    last_line = lines[-1]
+                with open(log_file_path, 'r') as log_file:
+                    line_list = log_file.read().splitlines()
```
Comment (tofu-rocketry) [inline check/check_db_dump_recent.py:38]: Indirectly relevant to this PR, but does this log get rotated? Have you noted the performance of this script?

File: check/check_db_dump_recent.py [hunk 6]
```diff
@@ -43,17 +45,30 @@ def run(self, log_file_path):
...
+                        "%Y-%m-%dT%H:%M:%S+0000",
+                    )
+                    # We only want the most recent success, so once we have
+                    # found it, break out of this loop
+                    break
+
+            if last_success is None:
+                print(
+                    "The failover process has never succeeded, "
+                    "according to the %s." % (
```
Comment (tofu-rocketry) [inline check/check_db_dump_recent.py:68]:
```suggestion
                    "according to %s." % (
```

The definite article sounds a bit weird before a file path.

File: test/test_check_db_dump_recent.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+happening.
+"""
+
+from datetime import datetime, timedelta
+import os
+import tempfile
+import unittest
+
+# Need to import the module as a whole to monkey patch the underlying
+# datetime now calls.
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:14]:
```suggestion
# datetime.now() calls.
```
To make it clear you're referring to a method.

File: test/test_check_db_dump_recent.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+import check.check_db_dump_recent
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:15]: Could have done an `import...as` to simplify the code, but prob not worth changing now.

File: test/test_check_db_dump_recent.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+        self.assertEqual(
+            return_code,
+            expected_return_code,
+            "Expected code %s, but got %s." % (
+                expected_return_code,
+                return_code,
+            ),
+        )
+
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:198]:
```suggestion


```

Two lines between classes.

File: test/test_check_db_dump_recent.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        self.assertEqual(
+            return_code,
+            check.check_db_dump_recent.RETURN_CODE_CRITICAL,
+            "Expected code %s, but got %s." % (
+                check.check_db_dump_recent.RETURN_CODE_CRITICAL,
+                return_code,
+            ),
+        )
+
+    def list_to_temp_file(self, input_list):
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:177]: Not marked as internal, but is this not another internal method like `_run_test`?

File: test/test_check_db_dump_recent.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        """Take the given list and write it to self.temp_file_path."""
+        with open(self.temp_file_path, "w") as temp_file:
+            for line in input_list:
+                temp_file.write(line + "\n")
+
+    def _run_test(self, log_line_list, expected_return_code):
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:183]: "_run_test" isn't the most descriptive of what it actually does, so there's an argument to put this nearer the top of the class before the `test_` methods.

File: test/test_check_db_dump_recent.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            "2023-01-01T10:40:01+0000",
+            "2023-01-01T10:41:01+0000 ERROR: An Error",
+            "2023-01-01T10:42:01+0000 ERROR: Another Error",
+            "2023-01-01T11:40:01+0000",
+            "2023-01-01T11:41:01+0000 ERROR: Yet another Error",
+        ]
+
+        self._run_test(log_line_list, check.check_db_dump_recent.RETURN_CODE_CRITICAL)
+
+    def test_recent_success_oracle(self):
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:96]: What's the difference between a mariadb and an oracle test?

**Review by tofu-rocketry — CHANGES_REQUESTED** [review]

File: importDBdmpFile/1_runDbUpdate.sh [hunk 13]
```diff
@@ -173,4 +173,5 @@ cd ..
...
 # Do not add any further "logger" statements after above line
```
Comment (gregcorbett) [inline importDBdmpFile/1_runDbUpdate.sh:177]:
```suggestion
# Do not remove following lines - other processes relies on these exact
# strings being the last line in the relevant log file.
echo "$(date --iso-8601='seconds') INFO: completed ok"
logger "completed ok"
# Do not add any further "logger" statements after above line
```

File: importDBdmpFile/1_runDbUpdate.sh (hunk 4 above)
Comment (gregcorbett) [reply importDBdmpFile/1_runDbUpdate.sh:176]: how about [this](https://github.com/GOCDB/gocdb-failover-scripts/pull/21/files/fd67bef3c44412b07ebaef954eab715d1cdd8837#r1339788706)?

File: test/test_check_db_dump_recent.py (hunk 12 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py:96]: The input, for Oracle error messages don't have a timestamp associated with them. For MariaDB they do.

File: check/check_db_dump_recent.py (hunk 5 above)
Comment (gregcorbett) [reply check/check_db_dump_recent.py:38]: The log does not currently get rotated. Extracting the whole file is not currently a performance issue.

File: test/test_check_db_dump_recent.py (hunk 8 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py:15]: while I'm updating other things, might as well address this as well 😄

Comment (gregcorbett) [reply test/test_check_db_dump_recent.py]: Done in 71218a77f13360e5a068a8f4de9b138c03ef9998

File: test/test_check_db_dump_recent.py (hunk 10 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py]: Fixed in 75199552ead172e431cadc215de6daa462e7bca7

File: test/test_check_db_dump_recent.py (hunk 11 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py]: Moved in 5b94a2dde297a445037f0a7e194c372e6f282383, open to suggestions on the name 😄

File: importDBdmpFile/1_runDbUpdate.sh (hunk 13 above)
Comment (tofu-rocketry) [reply importDBdmpFile/1_runDbUpdate.sh:177]: Ah, yes that's better.

File: test/test_check_db_dump_recent.py (hunk 11 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py]: discussed in team meeting, we are okay with the name as it following moving the method.

File: test/test_check_db_dump_recent.py (hunk 12 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py:112]: Greg to add class documentation to tests explaining format of example log files.

File: importDBdmpFile/1_runDbUpdate.sh [hunk 14]
```diff
@@ -170,7 +170,8 @@ mv $dmpTarget $dmpTargetFN.dmp
 # and move back up a dir as before
 cd ..
 
-# Do not remove following line - another process relies on this exact string
-# being the last line in the log file
+# Do not remove following lines - other processes relies on these exact
```
Comment (tofu-rocketry) [inline importDBdmpFile/1_runDbUpdate.sh:173]:
```suggestion
# Do not remove following lines - other processes rely on these exact
```

File: test/test_check_db_dump_recent.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+"""
+
+from datetime import datetime, timedelta
+import os
+import tempfile
+import unittest
+
+# Need to import the module as a whole to monkey patch the underlying
+# datetime.now() calls.
+import check.check_db_dump_recent as check_db_dump_recent
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:15]:
```suggestion
from check import check_db_dump_recent
```
Pretty sure you can do `from package import module` which would be more standard Python.

**Review by tofu-rocketry — CHANGES_REQUESTED** [review]

File: test/test_check_db_dump_recent.py (hunk 15 above)
Comment (tofu-rocketry) [reply test/test_check_db_dump_recent.py:15]: Well, not so much standard, but clearer as you're not actually renaming the module, just shortening calls to it.

Comment (gregcorbett) [reply test/test_check_db_dump_recent.py]: Addressed in 0ff8f3c515c42c4e75aeb6803d42e1ec24f5e568

File: test/test_check_db_dump_recent.py (hunk 12 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py:155]: Done in 529bdf35939aafeb68acc2b36c1d4c9d2d12d857

File: test/test_check_db_dump_recent.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+        )
+
+        # Create and close a temporary log file for use in the tests.
+        temp_file, self.temp_file_path = tempfile.mkstemp()
+        os.close(temp_file)
+
+        # Monkey patch the datetime module within check.check_db_dump_recent
+        # so we don't have to worry about hard coded datetimes in these tests.
+        check_db_dump_recent.datetime = MonkeyDateTime
+        help(check_db_dump_recent)
```
Comment (tofu-rocketry) [inline test/test_check_db_dump_recent.py:70]: What's the point of invoking `help` here?

**Review by tofu-rocketry — APPROVED** [review]
Looks ok, though puzzled by use of `help`.

File: test/test_check_db_dump_recent.py (hunk 16 above)
Comment (gregcorbett) [reply test/test_check_db_dump_recent.py:70]: ha, no point, i'll remove.

**Comment by gregcorbett** [comment]
I've squashed the commits down a bit, but [GitHub](https://github.com/GOCDB/gocdb-failover-scripts/compare/d2d681a52d95d667eed2b1bef229838061c145aa..9610cfe86503123a739d2f0334e479e4da81f8ce) confirms no changes have snuck in.

**Comment by tofu-rocketry** [comment]
We're all green then. :shipit:

**Pull request merged**